db/version_set.cc

```diff
@@ -508,23 +508,38 @@ class LevelIterator final : public InternalIterator {
     assert(Valid());
     return file_iter_.key();
   }
+
   Slice value() const override {
     assert(Valid());
     return file_iter_.value();
   }
+
   Status status() const override {
     return file_iter_.iter() ? file_iter_.status() : Status::OK();
   }
+
+  bool HintWithinLowerBound() override {
+    assert(Valid());
+    return hint_within_lower_bound_;
+  }
+
+  bool HintWithinUpperBound() override {
+    assert(Valid());
+    return file_iter_.HintWithinUpperBound();
```

Comment:
It's very complicated to reason about. How much impact is the block-based table reader part of the change? If we only do the hint for file boundary keys, will it achieve roughly the same?

Also, I don't understand why upper bound hint is pushed down to the file, rather than using the largest key of the file.

Reply:
For the reverse scan I only do file level optimize, and its result can be taken as a reference. We are getting 2% improvement for reverse vs. 5% for forward. The hint is push down to skip per-block key comparison if applicable. I'm seeing sub-1% improvement for doing it. The table reader currently don't know its largest key without reading data. Even the last index key can be substantial larger than the actual largest key.